pcdet/datasets/augmentor/augmentor_utils.py

```diff
@@ -76,43 +72,372 @@ def global_scaling(gt_boxes, points, scale_range):
+                np.logical_and(points[:,0] >= x - dx/2, \
+                    np.logical_and(points[:,2] <= z + dz/2, points[:,2] >= threshold))))))]
+
+    return gt_boxes, points
+
+def local_frustum_dropout_bottom(gt_boxes, points, intensity_range):
+    """
+    Args:
+        gt_boxes: (N, 7), [x, y, z, dx, dy, dz, heading, [vx], [vy]],
+        points: (M, 3 + C),
+        intensity: [min, max]
+    Returns:
+    """
+    for box in gt_boxes:
+        x, y, z, dx, dy, dz = box[0], box[1], box[2], box[3], box[4], box[5]
+
+        intensity = np.random.uniform(intensity_range[0], intensity_range[1])
+        get_points_in_box(points, box)
+        threshold = (z - dz/2) + intensity * dz
+        points = points[np.logical_not( \
+            np.logical_and(points[:,1] <= y + dy/2, \
+            np.logical_and(points[:,1] >= y - dy/2, \
+                np.logical_and(points[:,0] <= x + dx/2, \
+                    np.logical_and(points[:,0] >= x - dx/2, \
+                        np.logical_and(points[:,2] <= threshold, points[:,2] >= z - dz/2))))))]
+
+    return gt_boxes, points
+
+def local_frustum_dropout_left(gt_boxes, points, intensity_range):
+    """
+    Args:
+        gt_boxes: (N, 7), [x, y, z, dx, dy, dz, heading, [vx], [vy]],
+        points: (M, 3 + C),
+        intensity: [min, max]
+    Returns:
+    """
+    for box in gt_boxes:
+        x, y, z, dx, dy, dz = box[0], box[1], box[2], box[3], box[4], box[5]
+
+        intensity = np.random.uniform(intensity_range[0], intensity_range[1])
+        get_points_in_box(points, box)
+        threshold = (y + dy/2) - intensity * dy
+
+        points = points[np.logical_not( \
+            np.logical_and(points[:,1] <= y + dy/2, \
+            np.logical_and(points[:,1] >= threshold, \
+                np.logical_and(points[:,0] <= x + dx/2, \
+                    np.logical_and(points[:,0] >= x - dx/2, \
+                        np.logical_and(points[:,2] <= z + dz/2, points[:,2] >= z - dz/2))))))]
+
+    return gt_boxes, points
+
+def local_frustum_dropout_right(gt_boxes, points, intensity_range):
+    """
+    Args:
+        gt_boxes: (N, 7), [x, y, z, dx, dy, dz, heading, [vx], [vy]],
+        points: (M, 3 + C),
+        intensity: [min, max]
+    Returns:
+    """
+    for box in gt_boxes:
+        x, y, z, dx, dy, dz = box[0], box[1], box[2], box[3], box[4], box[5]
+
+        intensity = np.random.uniform(intensity_range[0], intensity_range[1])
+        get_points_in_box(points, box)
+        threshold = (y - dy/2) + intensity * dy
+
+        points = points[np.logical_not( \
+            np.logical_and(points[:,1] <= threshold, \
+            np.logical_and(points[:,1] >= y - dy/2, \
+                np.logical_and(points[:,0] <= x + dx/2, \
+                    np.logical_and(points[:,0] >= x - dx/2, \
+                        np.logical_and(points[:,2] <= z + dz/2, points[:,2] >= z - dz/2))))))]
+
+    return gt_boxes, points
+
+def get_points_in_box(points, gt_box):
```

Comment:
The results were visually validated using this tool: https://github.com/xaviersantos/OpenPCDet/blob/dev/data_augmentation.ipynb

Here are some images taken from these augmentations (the magnitude of the transformations was exaggerated for better visualization):

Original image:
![original](https://user-images.githubusercontent.com/33664182/143601679-2bf1cc67-19c5-4f55-94fb-904bce361447.png)

Local Translation along Y:
![translation_local_y](https://user-images.githubusercontent.com/33664182/143600799-83186f77-7871-437e-b016-e7e250fc5175.png)

Local Rotation:
![rotated_local](https://user-images.githubusercontent.com/33664182/143600814-2f54158e-44cb-4583-9173-38deab9e8701.png)

Local Scaling:
![scaled_local](https://user-images.githubusercontent.com/33664182/143600831-54c07a51-04bc-465f-882b-9b089bc2f975.png)

Comparing the results of the new augmentation methods with the ones already implemented, these were the improvements:

![table](https://user-images.githubusercontent.com/33664182/143600268-0950b4eb-9562-4512-8f43-d436f1899735.png)

These models models were trained for 80 epochs keeping all the same default hyperparameters.

As for the orientation of the gt_boxes, it should have no influence on the selection of the points inside its bounds (function get_points_in_box), as the center of the box and its dimensions are enough to the describe all the points within the box.

Reply:
The performance looks good!

But I still do not understand why you do not need to consider the orientation. 
Simply checking the center+dimensions of boxes for selecting inside points should only work for boxes with heading_angle=0. 
What would happen when you conduct this function for a gt_box with heading_angle=pi/4?   I think the inside points could not be well selected by this function since the gt_box is oblique. 

Actually I have implemented a function to select inside points from a rotated box (https://github.com/open-mmlab/OpenPCDet/blob/master/pcdet/ops/roiaware_pool3d/src/roiaware_pool3d.cpp#L121-L140), could you comment on the difference between your function and my function?

Reply:
I see what you mean and you're right.

I've fixed the get_points_in_box function using a numpy logical mask based on your code, as well as the local frustum dropout functions which had the same mistake.

The results were validated visually in scenes with objects in oblique positions and everything seems to be working as intended. I couldn't, however, train a model with this new fix to see if there are further improvements in performance yet, because the workstation I've been using for training is currently unavailable.

Reply:
Great! Thank you! 

I will try to train with your config and get back to here soon.

Reply:
@xaviersantos I find a function  `points_in_boxes_cpu` implemented in `./pcdet/ops/roiaware_pool3d/roiaware_pool3d_utils.py` similar with your vesion `get_points_in_box`. I would like to know why you don't use it.